File: scripts/check-indentation.py
#!/usr/bin/env python3
"""
check-indentation.py

Checks that all fenced code blocks in markdown/mdx files use 2-space indentation,
not 4-space or tab indentation.

A code block is flagged if:
  - Every indented line has indentation that is a strict multiple of 4 (pure 4-space block)
  - Or any line uses a tab character

Exit code 0 = all clear
Exit code 1 = violations found

Usage:
  python scripts/check-indentation.py
"""

import os
import sys

FENCE = "```"
DOCS_DIR = os.path.join(os.path.dirname(__file__), "..", "docs", "docs")


def check_file(fpath):
    violations = []
    with open(fpath, "r", encoding="utf-8") as f:
        lines = f.readlines()

    in_code_block = False
    code_block_lines = []
    block_start_line = 0
    lang = ""

    for i, line in enumerate(lines, start=1):
        stripped = line.rstrip("\n")

        if stripped.startswith(FENCE):
            if not in_code_block:
                in_code_block = True
                code_block_lines = []
                block_start_line = i + 1
                lang = stripped[3:].split()[0] if len(stripped) > 3 else ""
            else:
                violations += check_block(fpath, block_start_line, lang, code_block_lines)
                in_code_block = False
                code_block_lines = []
        elif in_code_block:
            code_block_lines.append((i, stripped))

    return violations


def check_block(fpath, block_start, lang, lines):
    violations = []
    tab_lines = [(lineno, l) for lineno, l in lines if l.startswith("\t")]
    indented = [(lineno, l) for lineno, l in lines if l.strip() and (len(l) - len(l.lstrip(" "))) > 0]

    # Flag tab indentation
    for lineno, l in tab_lines:
        violations.append({
            "file": fpath,
            "line": lineno,
            "block_start": block_start,
            "lang": lang,
            "reason": "tab indentation",
            "content": l[:60],
        })

    if not indented or tab_lines:
        return violations

    indent_levels = {len(l) - len(l.lstrip(" ")) for _, l in indented}

    # Flag if all indented lines are pure multiples of 4 (4-space block)
    if all(n % 4 == 0 for n in indent_levels):
        violations.append({
            "file": fpath,
            "line": block_start,
            "block_start": block_start,
            "lang": lang,
            "reason": f"4-space indentation (indent levels: {sorted(indent_levels)})",
            "content": None,
        })

    return violations


def main():
    docs_dir = os.path.abspath(DOCS_DIR)
    all_violations = []

    for root, dirs, files in os.walk(docs_dir):
        for fname in sorted(files):
            if fname.endswith(".md") or fname.endswith(".mdx"):
                fpath = os.path.join(root, fname)
                all_violations += check_file(fpath)

    if not all_violations:
        print("✅ All code blocks use correct 2-space indentation.")
        return 0

    print(f"❌ Found {len(all_violations)} indentation violation(s):\n")
    current_file = None
    for v in all_violations:
        rel = os.path.relpath(v["file"], docs_dir)
        if rel != current_file:
            print(f"  {rel}")
            current_file = rel
        if v["content"]:
            print(f"    line {v['line']:4d} ({v['lang'] or 'no lang'}): {v['reason']}")
            print(f"           {repr(v['content'])}")
        else:
            print(f"    line {v['line']:4d} ({v['lang'] or 'no lang'}): {v['reason']}")

    return 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/check-indentation.mjs
#!/usr/bin/env node
/**
 * check-indentation.mjs
 *
 * Checks that all fenced code blocks in markdown/mdx files use 2-space indentation,
 * not 4-space or tab indentation.
 *
 * A code block is flagged if:
 *   - Any line uses a tab character
 *   - Every indented line has indentation that is a strict multiple of 4 (pure 4-space block)
 *
 * Exit code 0 = all clear
 * Exit code 1 = violations found
 *
 * Usage:
 *   node scripts/check-indentation.mjs
 */

import { readFileSync, readdirSync, statSync } from 'fs';
import { join, resolve, relative, extname, dirname } from 'path';
import { fileURLToPath } from 'url';

const FENCE = '```';
const __dirname = dirname(fileURLToPath(import.meta.url));
const DOCS_DIR = resolve(__dirname, '..', 'docs', 'docs');

/**
 * @param {string} fpath
 * @param {number} blockStart
 * @param {string} lang
 * @param {Array<[number, string]>} lines
 * @returns {Array<object>}
 */
function checkBlock(fpath, blockStart, lang, lines) {
  const violations = [];
  const tabLines = lines.filter(([, l]) => l.startsWith('\t'));
  const indented = lines.filter(([, l]) => l.trim() && l.length - l.trimStart().length > 0);

  for (const [lineno, l] of tabLines) {
    violations.push({ file: fpath, line: lineno, lang, reason: 'tab indentation', content: l.slice(0, 60) });
  }

  if (!indented.length || tabLines.length > 0) return violations;

  const indentLevels = new Set(indented.map(([, l]) => l.length - l.trimStart().length));
  const sortedLevels = [...indentLevels].sort((a, b) => a - b);

  if (sortedLevels.every(n => n % 4 === 0)) {
    violations.push({
      file: fpath,
      line: blockStart,
      lang,
      reason: `4-space indentation (indent levels: ${sortedLevels})`,
      content: null,
    });
  }

  return violations;
}

/**
 * @param {string} fpath
 * @returns {Array<object>}
 */
function checkFile(fpath) {
  const violations = [];
  const lines = readFileSync(fpath, 'utf8').split('\n');

  let inCodeBlock = false;
  let codeBlockLines = [];
  let blockStartLine = 0;
  let lang = '';

  for (const [i, rawLine] of lines.entries()) {
    const lineno = i + 1;
    const stripped = rawLine.replace(/\r$/, '');

    if (stripped.startsWith(FENCE)) {
      if (!inCodeBlock) {
        inCodeBlock = true;
        codeBlockLines = [];
        blockStartLine = lineno + 1;
        lang = stripped.length > 3 ? stripped.slice(3).trim().split(/\s+/)[0] : '';
      } else {
        violations.push(...checkBlock(fpath, blockStartLine, lang, codeBlockLines));
        inCodeBlock = false;
        codeBlockLines = [];
      }
    } else if (inCodeBlock) {
      codeBlockLines.push([lineno, stripped]);
    }
  }

  return violations;
}

/**
 * Recursively collect all .md and .mdx files in a directory, sorted.
 *
 * @param {string} dir
 * @returns {string[]}
 */
function walkDir(dir) {
  const files = [];
  for (const entry of readdirSync(dir).sort()) {
    const full = join(dir, entry);
    if (statSync(full).isDirectory()) {
      files.push(...walkDir(full));
    } else if (extname(entry) === '.md' || extname(entry) === '.mdx') {
      files.push(full);
    }
  }
  return files;
}

const allViolations = [];

for (const fpath of walkDir(DOCS_DIR)) {
  allViolations.push(...checkFile(fpath));
}

if (!allViolations.length) {
  console.log('✅ All code blocks use correct 2-space indentation.');
  process.exit(0);
}

console.error(`❌ Found ${allViolations.length} indentation violation(s):\n`);

let currentFile = null;
for (const v of allViolations) {
  const rel = relative(DOCS_DIR, v.file);
  if (rel !== currentFile) {
    console.error(`  ${rel}`);
    currentFile = rel;
  }
  console.error(`    line ${String(v.line).padStart(4)} (${v.lang || 'no lang'}): ${v.reason}`);
  if (v.content !== null) {
    console.error(`           ${JSON.stringify(v.content)}`);
  }
}

process.exit(1);
